File: desktop/src/main/java/com/securetransfer/desktop/network/DeviceDiscovery.java
package com.securetransfer.desktop.network;

import com.securetransfer.desktop.model.Device;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.jmdns.JmDNS;
import javax.jmdns.ServiceEvent;
import javax.jmdns.ServiceInfo;
import javax.jmdns.ServiceListener;
import java.io.IOException;
import java.net.InetAddress;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

public class DeviceDiscovery {
    private static final Logger logger = LoggerFactory.getLogger(DeviceDiscovery.class);
    private static final String SERVICE_TYPE = "_securetransfer._tcp.local.";
    
    private JmDNS jmdns;
    private ServiceInfo serviceInfo;
    private final Map<String, Device> discoveredDevices = new ConcurrentHashMap<>();
    private final CopyOnWriteArrayList<Consumer<Device>> deviceAddedListeners = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<Consumer<Device>> deviceRemovedListeners = new CopyOnWriteArrayList<>();
    
    public void start() throws IOException {
        logger.info("Starting mDNS device discovery");
        
        InetAddress addr = InetAddress.getLocalHost();
        jmdns = JmDNS.create(addr);
        
        String deviceName = System.getProperty("user.name") + "'s " + 
                           System.getProperty("os.name").split(" ")[0];
        String deviceId = SecurityUtils.generateDeviceId();
        
        serviceInfo = ServiceInfo.create(SERVICE_TYPE, deviceName, 8080, 
            "deviceId=" + deviceId + ",platform=desktop");
        jmdns.registerService(serviceInfo);
        
        jmdns.addServiceListener(SERVICE_TYPE, new ServiceListener() {
            @Override
            public void serviceAdded(ServiceEvent event) {
                logger.debug("Service added: {}", event.getInfo().getName());
                jmdns.requestServiceInfo(event.getType(), event.getName(), 1000);
            }
            
            @Override
            public void serviceRemoved(ServiceEvent event) {
                logger.debug("Service removed: {}", event.getInfo().getName());
                String deviceId = event.getInfo().getPropertyString("deviceId");
                if (deviceId != null) {
                    Device device = discoveredDevices.remove(deviceId);
                    if (device != null) {
                        notifyDeviceRemoved(device);
                    }
                }
            }
            
            @Override
            public void serviceResolved(ServiceEvent event) {
                ServiceInfo info = event.getInfo();
                logger.debug("Service resolved: {}", info.getName());
                
                String deviceId = info.getPropertyString("deviceId");
                String platform = info.getPropertyString("platform");
                
                if (deviceId != null && !deviceId.equals(SecurityUtils.getLocalDeviceId())) {
                    Device device = new Device(
                        deviceId,
                        info.getName(),
                        "desktop".equals(platform) ? Device.Type.DESKTOP : Device.Type.MOBILE,
                        info.getInetAddresses()[0].getHostAddress(),
                        info.getPort(),
                        Device.Status.ONLINE,
                        System.currentTimeMillis()
                    );
                    
                    discoveredDevices.put(deviceId, device);
                    notifyDeviceAdded(device);
                }
            }
        });
        
        logger.info("mDNS service started on {}", addr.getHostAddress());
    }
    
    public void stop() {
        logger.info("Stopping mDNS device discovery");
        
        if (jmdns != null) {
            try {
                if (serviceInfo != null) {
                    jmdns.unregisterService(serviceInfo);
                }
                jmdns.close();
            } catch (IOException e) {
                logger.error("Error stopping mDNS service", e);
            }
        }
    }
    
    public void addDeviceAddedListener(Consumer<Device> listener) {
        deviceAddedListeners.add(listener);
    }
    
    public void addDeviceRemovedListener(Consumer<Device> listener) {
        deviceRemovedListeners.add(listener);
    }
    
    private void notifyDeviceAdded(Device device) {
        logger.info("Device discovered: {} at {}:{}", device.getName(), 
                   device.getIpAddress(), device.getPort());
        deviceAddedListeners.forEach(listener -> {
            try {
                listener.accept(device);
            } catch (Exception e) {
                logger.error("Error notifying device added listener", e);
            }
        });
    }
    
    private void notifyDeviceRemoved(Device device) {
        logger.info("Device removed: {}", device.getName());
        deviceRemovedListeners.forEach(listener -> {
            try {
                listener.accept(device);
            } catch (Exception e) {
                logger.error("Error notifying device removed listener", e);
            }
        });
    }
    
    public Map<String, Device> getDiscoveredDevices() {
        return new ConcurrentHashMap<>(discoveredDevices);
    }
}

File: android/src/main/java/com/securetransfer/android/network/DeviceDiscovery.kt
package com.securetransfer.android.network

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import com.securetransfer.android.model.Device
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.slf4j.LoggerFactory
import java.util.concurrent.ConcurrentHashMap

class DeviceDiscovery(private val context: Context) {
    private val logger = LoggerFactory.getLogger(DeviceDiscovery::class.java)
    
    companion object {
        private const val SERVICE_TYPE = "_securetransfer._tcp"
        private const val SERVICE_NAME = "SecureTransfer"
    }
    
    private val nsdManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    private val _discoveredDevices = MutableStateFlow<Map<String, Device>>(emptyMap())
    val discoveredDevices: StateFlow<Map<String, Device>> = _discoveredDevices
    
    private val devices = ConcurrentHashMap<String, Device>()
    private var isDiscovering = false
    private var isRegistered = false
    
    private val discoveryListener = object : NsdManager.DiscoveryListener {
        override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
            logger.error("Discovery start failed: $errorCode")
        }
        
        override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
            logger.error("Discovery stop failed: $errorCode")
        }
        
        override fun onDiscoveryStarted(serviceType: String) {
            logger.info("Discovery started for: $serviceType")
            isDiscovering = true
        }
        
        override fun onDiscoveryStopped(serviceType: String) {
            logger.info("Discovery stopped for: $serviceType")
            isDiscovering = false
        }
        
        override fun onServiceFound(serviceInfo: NsdServiceInfo) {
            logger.debug("Service found: ${serviceInfo.serviceName}")
            if (serviceInfo.serviceType == SERVICE_TYPE && 
                serviceInfo.serviceName != SERVICE_NAME) {
                nsdManager.resolveService(serviceInfo, resolveListener)
            }
        }
        
        override fun onServiceLost(serviceInfo: NsdServiceInfo) {
            logger.debug("Service lost: ${serviceInfo.serviceName}")
            val deviceId = serviceInfo.serviceName
            devices.remove(deviceId)?.let {
                _discoveredDevices.value = devices.toMap()
            }
        }
    }
    
    private val resolveListener = object : NsdManager.ResolveListener {
        override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
            logger.error("Resolve failed for ${serviceInfo.serviceName}: $errorCode")
        }
        
        override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
            logger.debug("Service resolved: ${serviceInfo.serviceName}")
            
            val deviceId = serviceInfo.serviceName
            val host = serviceInfo.host?.hostAddress ?: return
            val port = serviceInfo.port
            
            val device = Device(
                id = deviceId,
                name = serviceInfo.serviceName,
                type = Device.Type.DESKTOP, // Will be updated based on service attributes
                ipAddress = host,
                port = port,
                status = Device.Status.ONLINE,
                lastSeen = System.currentTimeMillis()
            )
            
            devices[deviceId] = device
            _discoveredDevices.value = devices.toMap()
            
            logger.info("Device discovered: ${device.name} at $host:$port")
        }
    }
    
    private val registrationListener = object : NsdManager.RegistrationListener {
        override fun onRegistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
            logger.error("Registration failed: $errorCode")
        }
        
        override fun onUnregistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
            logger.error("Unregistration failed: $errorCode")
        }
        
        override fun onServiceRegistered(serviceInfo: NsdServiceInfo) {
            logger.info("Service registered: ${serviceInfo.serviceName}")
            isRegistered = true
        }
        
        override fun onServiceUnregistered(serviceInfo: NsdServiceInfo) {
            logger.info("Service unregistered: ${serviceInfo.serviceName}")
            isRegistered = false
        }
    }
    
    fun startDiscovery() {
        if (!isDiscovering) {
            logger.info("Starting device discovery")
            nsdManager.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, discoveryListener)
        }
    }
    
    fun stopDiscovery() {
        if (isDiscovering) {
            logger.info("Stopping device discovery")
            nsdManager.stopServiceDiscovery(discoveryListener)
        }
    }
    
    fun registerService(port: Int) {
        if (!isRegistered) {
            logger.info("Registering service on port $port")
            
            val serviceInfo = NsdServiceInfo().apply {
                serviceName = SERVICE_NAME
                serviceType = SERVICE_TYPE
                this.port = port
            }
            
            nsdManager.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, registrationListener)
        }
    }
    
    fun unregisterService() {
        if (isRegistered) {
            logger.info("Unregistering service")
            // Note: We need to keep a reference to the service info used for registration
            // This is a simplified version - in practice, you'd store the registered service info
        }
    }
    
    fun cleanup() {
        stopDiscovery()
        unregisterService()
        devices.clear()
        _discoveredDevices.value = emptyMap()
    }
}
